plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/IsNullPushdownDataTypeTest.java

```diff
@@ -0,0 +1,211 @@
+
+        try {
+            assertPushdown(expectPushdown,
+                    assertResult(scenario.equals(IS_NULL) ? Optional.of(firstCase.expectedLiteral()) : Optional.empty(),
+                            assertThat(queryAssertions.query(session, queryWithAll))));
+        }
+        catch (AssertionError e) {
+            for (int column = SPECIAL_COLUMNS; column < testCases.size(); column++) {
+                String queryWithSingleColumnPredicate = "SELECT " + firstColumnName + " FROM " + temporaryRelation.getName() + " WHERE " + getPredicate(column, scenario) + withConnectorExpression;
+                assertPushdown(expectPushdown,
+                        assertResult(scenario.equals(IS_NULL) ? Optional.of(firstCase.expectedLiteral()) : Optional.empty(),
+                                assertThat(queryAssertions.query(session, queryWithSingleColumnPredicate))));
+            }
+            throw new IllegalStateException("Single column assertion should fail for at least one column, if query of all column failed", e);
+        }
+    }
+
+    private String getPredicate(int column, Scenario scenario)
+    {
+        checkArgument(column >= SPECIAL_COLUMNS);
+        String columnName = "col_" + column;
+        checkArgument("NULL".equalsIgnoreCase(testCases.get(column).inputLiteral()));
+        return scenario == IS_NULL
+                ? columnName + " IS NULL"
+                : columnName + " IS NOT NULL";
+    }
+
+    private static QueryAssertions.QueryAssert assertResult(Optional<String> value, QueryAssertions.QueryAssert assertion)
+    {
+        return value.isPresent()
+                ? assertion.matches("VALUES %s".formatted(value.get()))
+                : assertion.returnsEmptyResult();
+    }
+
+    private static QueryAssertions.QueryAssert assertPushdown(boolean expectPushdown, QueryAssertions.QueryAssert assertion)
+    {
+        return expectPushdown
+                ? assertion.isFullyPushedDown()
+                : assertion.isNotFullyPushedDown(FilterNode.class);
+    }
+
+    private record TestCase(
+            Optional<String> declaredType,
+            String inputLiteral,
+            Optional<Type> expectedType,
+            String expectedLiteral)
+            implements ColumnSetup
+    {
+        private TestCase
+        {
+            requireNonNull(declaredType, "declaredType is null");
+            requireNonNull(expectedType, "expectedType is null");
+            requireNonNull(inputLiteral, "inputLiteral is null");
+            requireNonNull(expectedLiteral, "expectedLiteral is null");
+        }
+
+        @Override
+        public Optional<String> getDeclaredType()
+        {
+            return declaredType;
+        }
+
+        @Override
+        public String getInputLiteral()
+        {
+            return inputLiteral;
+        }
+    }
+
+    enum Scenario
+    {
+        IS_NULL,
+        IS_NOT_NULL,
+    }
+
+    public enum PushdownImplementation
+    {
+        NEVER,
```

Comment:
unused

Reply:
However it gives a sense about possible values.